drivers/storage/portworx/component/plugin.go

```diff
@@ -0,0 +1,359 @@
+	if err := p.createDeployment(pluginDeploymentFileName, cluster.Namespace, ownRef); err != nil {
+		errList = append(errList, err.Error())
+		logrus.Errorf("error during creating deployment %s ", err)
+	}
+	if err := p.createService(pluginServiceFileName, cluster.Namespace, ownRef); err != nil {
+		errList = append(errList, err.Error())
+		logrus.Errorf("error during creating service %s ", err)
+	}
+	if err := p.createConfigmap(pluginConfigmapFileName, cluster.Namespace, ownRef); err != nil {
+		errList = append(errList, err.Error())
+		logrus.Errorf("error during creating config map  %s ", err)
+	}
+
+	//create console plugin
+	if err := p.createConsolePlugin(consolePluginFileName, cluster.Namespace, ownRef); err != nil {
+		errList = append(errList, err.Error())
+		logrus.Errorf("error during creating console plugin %s ", err)
+	}
+
+	if len(errList) > 0 {
+		return commonerrors.New(strings.Join(errList, " , "))
+	}
+	return nil
+}
+
+func (p *plugin) Delete(cluster *corev1.StorageCluster) error {
+	var errList []string
+	ownerRef := metav1.NewControllerRef(cluster, pxutil.StorageClusterKind())
+
+	//delete plugin configs
+	if err := k8s.DeleteDeployment(p.client, PluginDeloymentName, cluster.Namespace, *ownerRef); err != nil {
+		errList = append(errList, err.Error())
+	}
+	if err := k8s.DeleteConfigMap(p.client, PluginConfigMapName, cluster.Namespace, *ownerRef); err != nil {
+		errList = append(errList, err.Error())
+	}
+	if err := k8s.DeleteService(p.client, PluginServiceName, cluster.Namespace, *ownerRef); err != nil {
+		errList = append(errList, err.Error())
+	}
+
+	//delete nginx configs
+	if err := k8s.DeleteDeployment(p.client, NginxDeploymentName, cluster.Namespace, *ownerRef); err != nil {
+		errList = append(errList, err.Error())
+	}
+	if err := k8s.DeleteConfigMap(p.client, NginxConfigMapName, cluster.Namespace, *ownerRef); err != nil {
+		errList = append(errList, err.Error())
+	}
+	if err := k8s.DeleteService(p.client, NginxServiceName, cluster.Namespace, *ownerRef); err != nil {
+		errList = append(errList, err.Error())
+	}
+
+	//delete console plugin
+	if err := p.deleteConsolePlugin(PluginName, cluster.Namespace, *ownerRef); err != nil {
+		errList = append(errList, err.Error())
+	}
+	if len(errList) > 0 {
+		return commonerrors.New(strings.Join(errList, " , "))
+	}
+	return nil
+}
+
+func (c *plugin) MarkDeleted() {}
+
+// RegisterPortworxPluginComponent registers the PortworxPlugin component
+func RegisterPortworxPluginComponent() {
+	Register(PluginComponentName, &plugin{})
+}
+
+func init() {
+	RegisterPortworxPluginComponent()
+}
+
+func (p *plugin) createDeployment(filename, storageNs string, ownerRef *metav1.OwnerReference) error {
+	deployment, err := k8s.GetDeploymentFromFile(filename, pxutil.SpecsBaseDir())
+	if err != nil {
+		return err
+	}
+	deployment.Namespace = storageNs
+	deployment.OwnerReferences = []metav1.OwnerReference{*ownerRef}
+	return k8s.CreateOrUpdateDeployment(p.client, deployment, ownerRef)
```

Comment:
Unfortunately the CreateOrUpdateDeployment does not work like the other funcs for other objects. As k8s adds some default values in the spec, the diff between what is in k8s api server and what we read from file is always non-nil. So we end up calling api server to update the deployment every time.

To avoid this we introduced a very crude way of keeping track of the deployment using flags. Check other components on how we deal with deployments.


Also, we should add extra things to the deployment like the image registry, image pull secret, toleration, etc. Again, check other components.